pkg/addon/addon_update.go

```diff
@@ -0,0 +1,320 @@
+				msg := fmt.Sprintf("[updateAddon] Addon %s/%s is not found. %v", updated.Namespace, updated.Name, err)
+				//c.logger.Error(msg)
+				return nil, fmt.Errorf(msg)
+			case strings.Contains(err.Error(), "the object has been modified"):
+				errs = append(errs, err)
+				c.log.Info(fmt.Sprintf("[updateAddon] retry updating object metadata %s/%s coz objects has been modified", updated.Namespace, updated.Name))
+				if _, err := c.updateAddon(ctx, updated); err != nil {
+					errs = append(errs, err)
+				}
+			default:
+				return nil, err
+			}
+		}
+		_, err = c.updateAddonStatus(ctx, updated)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) == 0 {
+		return updated, nil
+	}
+	c.log.Error(err, fmt.Sprintf("[updateAddon] %s/%s failed.", updated.Namespace, updated.Name))
+	return nil, fmt.Errorf("%v", errs)
+}
+
+func (c *AddonUpdate) getExistingAddon(ctx context.Context, key string) (*addonv1.Addon, error) {
+	info := strings.Split(key, "/")
+	updating := &addonv1.Addon{}
+	err := c.client.Get(ctx, types.NamespacedName{Namespace: info[0], Name: info[1]}, updating)
+	if err != nil || updating == nil {
+		return nil, fmt.Errorf("[getExistingAddon] failed converting to addon %s from client err %#v", key, err)
+	}
+	return updating, nil
+}
+
+func (c *AddonUpdate) addAddonToCache(addon *addonv1.Addon) {
+	var version = Version{
+		Name:        addon.GetName(),
+		Namespace:   addon.GetNamespace(),
+		PackageSpec: addon.GetPackageSpec(),
+		PkgPhase:    addon.GetInstallStatus(),
+	}
+	c.versionCache.AddVersion(version)
+	c.log.WithValues("addAddonToCache", fmt.Sprintf("Adding %s/%s package %s version cache phase %s", addon.GetNamespace(), addon.GetName(), addon.GetPackageSpec(), version.PkgPhase))
+}
+
+// UpdateAddonStatusLifecycle
+func (c *AddonUpdate) UpdateAddonStatusLifecycle(ctx context.Context, namespace, name string, lifecycle string, lifecyclestatus wfv1.WorkflowPhase) error {
+
+	key := fmt.Sprintf("%s/%s", namespace, name)
+	latest, err := c.getExistingAddon(ctx, key)
+	if err != nil || latest == nil {
+		return err
+	}
+	updating := latest.DeepCopy()
+	prevStatus := latest.Status
+
+	// addon being deletion, skip non-delete wf update
+	if lifecycle != "delete" &&
+		prevStatus.Lifecycle.Installed == addonv1.Deleting {
+		return nil
+	}
+
+	newStatus := addonv1.AddonStatus{
+		Lifecycle: addonv1.AddonStatusLifecycle{},
+		Resources: []addonv1.ObjectStatus{},
+	}
+	newStatus.Reason = prevStatus.Reason
+	newStatus.Resources = append(newStatus.Resources, prevStatus.Resources...)
+	newStatus.Checksum = prevStatus.Checksum
+	newStatus.StartTime = prevStatus.StartTime
+	if lifecycle == "prereqs" {
+		newStatus.Lifecycle.Prereqs = addonv1.ApplicationAssemblyPhase(lifecyclestatus)
+		newStatus.Lifecycle.Installed = prevStatus.Lifecycle.Installed
+		if newStatus.Lifecycle.Prereqs == addonv1.Failed {
+			newStatus.Lifecycle.Installed = addonv1.Failed
+			newStatus.Reason = "prereqs wf fails."
```

Comment:
```suggestion
			newStatus.Reason = "prereqs wf failed."
```

Reply:
under this level, wf-name already thrown away, do not apply here anymore.

Reply:
Get rid of reason if it's already populated

Reply:
07903cafbefbde5e80a681630e88d2e6f75aace1